给ExternalCatalog增加partition的概念

beyond1920 opened the pull request:
在ExternalCatalog上增加partition的概念，包括：
1. 引入分区的Mode
2. 在ExternalCatalog上增加partition的增删改查接口
分为两个commit：
第一个commit对应https://github.com/beyond1920/flink/pull/2， 第二个commit是增加partition概念，review 这个pr只需要关注第二个commit即可。

chenqin committed:
[FLINK-4460] Add support for side outputs

This does not yet allow users to emit to side outputs in user functions.
Only operators (StreamOperator) can emit to side outputs. A side output
can be retrieved on a SingleOutputStreamOperator.

aljoscha committed:
[FLINK-4460] Add side outputs for ProcessFunction

tonycox committed:
[FLINK-3849] [table] Add FilterableTableSource interface and rules for pushing it (1)

fix filterable test

rebase and trying fix rexnode parsing

create wrapper and update rules

Commit:
[FLINK-4364] [heartbeats] Implement TaskManager side of heartbeat from JobManager

This closes #3151.

jgrier committed:
[FLINK-5635] [docker] Improvements for Docker on Flink experience

Modifying Dockerfile to build from local flink-dist as well as release URLs.
Logging to stdout.
Adding scripts to deploy seamlessly on Docker Swarm.
Updating Docker Compose scripts to work correctly.
Parameterizing things so these Docker scripts are more generally useful.

GEOFBOT committed:
[FLINK-5183] [py] Support mulitple jobs per plan file

This closes #3232.

aljoscha committed:
[FLINK-4460] Make chaining work with side outputs

aljoscha committed:
[FLINK-4460] Expose OutputTag constructor that takes TypeInformation

tony810430 committed:
[FLINK-4754] [checkpoints] Make number of retained checkpoints user configurable

This closes #3374

kl0u committed:
[FLINK-5846] [cep] Make the CEP operators backwards compatible with Flink 1.1

jinmingjian committed:
[FLINK-5692] [core] Add an Option to Deactivate Kryo Fallback for Serializers

This closes #3373

jinmingjian committed:
[FLINK-4422] [clients] Convert time interval measurements to System.nanoTime() in 'flink-clients'

This closes #3384

kl0u committed:
[FLINK-5932] initializeState() before legacy state restoring in operator

chenqin committed:
[FLINK-4460] Provide late-data output for window operations

We use side outputs to emit dropped late data.

Commit:
[FLINK-5134] [runtime] Aggregate ResourceSpec for chained operators when generating job graph

This closes #3455

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala:
```diff
@@ -40,6 +40,7 @@ case class ExternalCatalogTable(
     properties: Map[String, String] = Map.empty,
     stats: Option[TableStats] = None,
     comment: Option[String] = None,
+    partitionColumnNames: Seq[String] = Seq.empty,
```
partitionColumnNames 的类型建议为为 Option[Seq[String]] 。一张表可能不是 partition 表。 Seq.empty 表示 partition表中 目前没有 partition

beyond1920 replied:
我需要加一个字断表示是否是分区表，之前本来想通过partitionColumnNames是否为empty来说明是否是分区表，但是由于以下两个原因不行：
1. 可能是动态分区，这样确实有可能存在是分区表但是partitionColumnNames是empty
2. 是视图，比如odps view，这种视图的partitionColumnNames不是empty，但是也不是分区表

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTablePartition.scala:
```diff
@@ -0,0 +1,42 @@
+    * external table partition specification.
+    * Key is partition column name, value is partition column value.
+    */
+  type PartitionSpec = Map[String, String]
```
如何保证是有层级关系的 partition 的表示时的列的顺序

beyond1920 replied:
partitionColumns里面的是partition column的顺序

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/InMemoryExternalCatalog.scala:
```diff
@@ -120,14 +112,109 @@ class InMemoryExternalCatalog extends ExternalCatalog {
+
+  private def checkPartitionSpec(partSpec: PartitionSpec, table: ExternalCatalogTable): Unit =
+    if (partSpec.keySet != table.partitionColumnNames.toSet) {
+      throw new IllegalArgumentException("Input partition specification is invalid!")
```
显示提醒是分区表的错误还是非分区表的错误

beyond1920 replied:
？？这个地方实际是是说分区列有ds,hour，但是输入的partitionSpec可能是Map("ds" -> "20140706)，或者("ds"->"20140706", "time"=12); 这两种情况都不合法，
输入的map中的key必须和partitionColumns对应上

StefanRRichter committed:
[FLINK-5715] Asynchronous snapshots for heap-based keyed state backend

aljoscha committed:
[FLINK-4460] Add proper side output API for Scala API

The Scala side output API uses context bounds to get a TypeInformation
for an OutputTag. This also adds a SideOutputITCase for the Scala API.

aljoscha committed:
[FLINK-4460] Add documentation for side outputs

WangTaoTheTonic committed:
[FLINK-5981] [security] Make SSL pick up configured protocols and cipher suites

This closes #3486

patricklucas committed:
[FLINK-5635] [docker] Use start-foreground in Docker entrypoint

docker-entrypoint.sh should error on invalid args

Improve docker build.sh cleanup

Dockerfile improvements per review

Remove unnecessary Dockerfile build steps

Now that docker-entrypoint.sh uses 'start-foreground', munging
flink-daemon.sh and overwriting the log4j config are not longer
necessary.

Improve Dockerfile and docker-entrypoint.sh

Clean up Dockerfile and improve docker-entrypoint.sh to support '--help'
and pass through non-(jobmanager|taskmanager) commands.

This closes #3205.
This closes #3494.

tillrohrmann committed:
[FLINK-4364] Introduce HeartbeatServices for the HeartbeatManager instantiation

The HeartbeatServices are used to create all services relevant for heartbeating. This
includes at the moment the creation of HeartbeatManager implementations which actively
send heartbeats and those which only respond to heartbeat requests.

Add comments

greghogan committed:
[FLINK-6009] [java api] Deprecate DataSetUtils#checksumHashCode

This is likely only used by Gelly and we have a more featureful
implementation allowing for multiple outputs and setting the job name.
Deprecation will allow this to be removed in Flink 2.0.

This closes #3516

aljoscha committed:
[FLINK-4460] Add WindowOperatorContractTest tests for late data emission

aljoscha committed:
[FLINK-5808] Move default parallelism to StreamingJobGraphGenerator

Before, it was set on the ExecutionConfig for some stream execution
environments and later for others. Now, we don't set the default
parallelism on the ExecutionConfig but instead set it at the latest
possible point, in the StreamingJobGraphGenerator.

This also adds tests that verify that we don't set the default
parallelism on the ExecutionConfig.

aljoscha committed:
[FLINK-5808] Add proper checks in setParallelism()/setMaxParallelism()

Before, there where some checks in
StreamExecutionEnvironment.set(Max)Parallelism() but a user would
circumvent these if using the ExecutionConfig directly. Now, all checks
are moved to the ExecutionConfig.

aljoscha committed:
[FLINK-5808] Move max keygroup constants to ExecutionConfig

We need to have them there if we want to properly test the arguments of
setMaxParallelism() in the ExecutionConfig itself.

aljoscha committed:
[FLINK-4460] Add ITCase to verify watermark forwarding for side outputs

StefanRRichter committed:
[FLINK-5985] Report no task states for stateless tasks on checkpointing

aljoscha committed:
[FLINK-4460] Add side output ITCase for multiple side out consumers

kl0u committed:
[hotfix] [doc] Fix error in ProcessFunction example.

aljoscha committed:
[FLINK-4460] Catch side output OutputTag id clashes

This also adds tests.

KurtYoung committed:
[FLINK-3849] [table] Add FilterableTableSource interface and rules for pushing it (2)

This closes #3520.

fix compilation failure

fix compilation failure again.

1. Deep copy TableSource when we copy TableSourceScan
2. unify push project into scan rule for both batch and stream

address comments.

expand project list before creating new RexProgram

update tests

aljoscha committed:
[FLINK-6001] Fix ContinuousEventTimeTrigger firing without state

tillrohrmann committed:
[FLINK-6076] Refactor HeartbeatManager to extend HeartbeatTarget

Remove start method from HeartbeatManager

This closes #3555.

kl0u committed:
[FLINK-6032] [cep] Clean-up operator state when not needed.

The CEP operator now cleans the registered state for a
key. This happens:
1) for the priority queue, when the queue is empty.
2) for the NFA, when its shared buffer is empty.
3) finally the key is removed from the watermark
   callback service if both the above are empty.

liuyuzhong7 committed:
[FLINK-6041] [streaming api] Move StreamingFunctionUtils to 'org.apache.flink.streaming.util'

This close #3532

StefanRRichter committed:
[FLINK-6044] Replace all unintentional calls to InputStream#read(...) with InputStream#readFully(...)

sunjincheng121 committed:
[FLINK-6018] Properly initialise StateDescriptor in AbstractStateBackend.getPartitionedState()

Commit:
[FLINK-6051] [metrics] [docs] Fix scope keys on Configuration page

This closes #3538.

StefanRRichter committed:
[FLINK-6061] [state backends] Throw IllegalStateException when using RocksDB keyed state with no key set

StephanEwen committed:
[FLINK-5692] [core] (followups) Add an Option to Deactivate Kryo Fallback for Serializers

StephanEwen committed:
[hotfix] [core] Fix/cleanup serialization test for ExecutionConfig

Commit:
[FLINK-6064][flip6] fix BlobServer connection in TaskExecutor

The hostname used for the BlobServer was set to the akka address which is
invalid for this use. Instead, this adds the hostname to the RpcGateway /
AkkaInvocationHandler so that this information is available to the TaskExecutor.

This closes #3551.

StephanEwen committed:
[FLINK-5962] [checkpoints] Remove scheduled cancel-task from timer queue to prevent memory leaks

This closes #3548

zentol committed:
[FLINK-5650] [py] Continuously check PyProcess health while waiting for incoming connection

StephanEwen committed:
[FLINK-5756] [rocksdb] Add mini benchmarks to reproduce 'merge' performance problems

zhuoluoy committed:
[FLINK-6068] [table] Support If() as a built-in function.

This cloes #3553.

StephanEwen committed:
[FLINK-4754] [checkpoints] Small followups to the configuration of number of retained checkpoints.

zentol committed:
[refactor] [py] General code clean-up

- added missing serialVersionUID
- PythonPlan* classes no longer implement serializable
- marked several fields as transient
- replaced some fields by local variables
- input/output files are now being deleted
- remvoed some unnecessary casts
- renamved several ignored exceptions
- close socket in PythonPlanStreamer
- simplified PyProcess health check
- eliminated several uses of raw types
- removed some unthrown exception declarations

aljoscha committed:
[hotfix] Remove auto-boxing in ContinuousEventTimeTrigger

StefanRRichter committed:
[hotfix] Added missing documentation in StateTransformationFunction

beyond1920 committed:
[FLINK-5568] [table] Introduce ExternalCatalog interface, integrated with calcite and provide an in-memory implementation for testing.

This closes #3406.

zentol committed:
[FLINK-6111] [py] Remove unnecessary sleeps

beyond1920 committed:
Let catalog support partition
